src/interned.rs

```diff
@@ -43,60 +45,136 @@ pub struct JarImpl<C: Configuration> {
+    lru: LinkedList<ValueAdapter<C>>,
+}
+
+impl<C: Configuration> Default for IngredientShard<C> {
+    fn default() -> Self {
+        Self {
+            lru: LinkedList::default(),
+            key_map: hashbrown::HashTable::new(),
+        }
+    }
+}
+
+// SAFETY: `LinkedListLink` is `!Sync`, however, the linked list is only accessed through the
+// ingredient lock, and values are only ever linked to a single list on the ingredient.
+unsafe impl<C: Configuration> Sync for Value<C> {}
+
+intrusive_adapter!(ValueAdapter<C> = UnsafeRef<Value<C>>: Value<C> { link: LinkedListLink } where C: Configuration);
+
 /// Struct storing the interned fields.
 pub struct Value<C>
 where
     C: Configuration,
 {
-    fields: C::Fields<'static>,
-    memos: MemoTable,
+    /// The index of the shard containing this value.
+    shard: u16,
 
-    /// The revision the value was first interned in.
-    first_interned_at: Revision,
+    /// An intrusive linked list for LRU.
+    link: LinkedListLink,
 
-    /// The most recent interned revision.
-    last_interned_at: AtomicRevision,
+    /// The interned fields for this value.
+    ///
+    /// These are valid for read-only access as long as the lock is held
+    /// or the value has been validated in the current revision.
+    fields: UnsafeCell<C::Fields<'static>>,
+
+    /// Memos attached to this interned value.
+    ///
+    /// This is valid for read-only access as long as the lock is held
+    /// or the value has been validated in the current revision.
+    memos: UnsafeCell<MemoTable>,
+
+    /// Data that can only be accessed while holding the lock for the
+    /// `key_map` shard containing the value ID.
+    shared: UnsafeCell<ValueShared>,
+}
+
+/// Shared value data can only be read through the lock.
+struct ValueShared {
+    /// The interned ID for this value.
+    ///
+    /// Storing this on the value itself is necessary to identify slots
+    /// from the LRU list, as well as keep track of the generation.
+    ///
+    /// Values that are reused increment the ID generation, as if they had
+    /// allocated a new slot. This eliminates the need for dependency edges
+    /// on queries that *read* from an interned value, as any memos dependent
+    /// on the previous value will not match the new ID.
+    ///
+    /// However, reusing a slot invalidates the previous ID, so dependency edges
+    /// on queries that *create* an interned value are still required to ensure
+    /// the value is re-interned with a new ID.
+    id: Id,
+
+    /// The revision the value was most-recently interned in.
+    last_interned_at: Revision,
 
     /// The minimum durability of all inputs consumed by the creator
-    /// query prior to creating this tracked struct. If any of those
+    /// query prior to creating this interned struct. If any of those
     /// inputs changes, then the creator query may create this struct
     /// with different values.
-    durability: AtomicU8,
+    durability: Durability,
```

Comment:
Note that this field wastes 8 bytes while only occupying a single byte due to padding